### PR TITLE
GCP effective firewall listing

### PR DESCRIPTION
**Pull request by jonathan-greig**
Relates to #339

### CODE REVIEW

File: tests/providers/gcp/gcp_mocks.py [hunk 1]
```diff
@@ -556,3 +556,97 @@
...
+            {
+                'type': 'ONE_TO_ONE_NAT',
+                'name': 'External NAT',
+                'natIP': '0.0.0.0',
+                'networkTier': 'PREMIUM',
+                'kind': 'compute#accessConfig'
+                }
+        ],
+        'fingerprint': 'bm9mcGZwZnA=',
+        'kind': 'compute#networkInterface'}
```
Comment (giovannt0) [inline tests/providers/gcp/gcp_mocks.py:577]: Nit: put the bracket on the next line so we clearly see where the dict finishes

File: libcloudforensics/providers/gcp/internal/compute.py [hunk 2]
```diff
@@ -1224,6 +1224,85 @@ def DetachServiceAccount(self) -> None:
...
+    gce_instance_client = self.GceApi().instances()
+    instance_info = self.GetOperation()
+    fw_rules = {}
+
+    for nic in instance_info.get('networkInterfaces', []):
+      nic_name = nic['name']
+      request = gce_instance_client.getEffectiveFirewalls(
+          project=self.project_id, instance=self.name, zone=self.zone,
+          networkInterface=nic_name)
+      response = request.execute()
```
Comment (giovannt0) [inline libcloudforensics/providers/gcp/internal/compute.py:1300]: You may wanna wrap this using `common.ExecuteRequest`, so that errors are caught.

**Review by giovannt0 — CHANGES_REQUESTED** [review]

**Comment by jonathan-greig** [comment]
Addressed code review comments and also did some further normalisation of the fw rules as there were inconsistencies in FW vs FW policy field names `IPProtocol` vs `ipProtocol`

File: libcloudforensics/providers/gcp/internal/compute.py [hunk 3]
```diff
@@ -1224,6 +1226,112 @@ def DetachServiceAccount(self) -> None:
...
+    gce_instance_client = self.GceApi().instances()
+    instance_info = self.GetOperation()
+    fw_rules = {}
+
+    for nic in instance_info.get('networkInterfaces', []):
+      nic_name = nic['name']
+      nic_fw_rules = []
+      responses = common.ExecuteRequest(
+          gce_instance_client, 'getEffectiveFirewalls',
+          {'project': self.project_id, 'instance': self.name,
```
Comment (giovannt0) [inline libcloudforensics/providers/gcp/internal/compute.py:1326]: Can you make this a separate variable? `request = {...}`

File: libcloudforensics/providers/gcp/internal/compute.py [hunk 4]
```diff
@@ -1224,6 +1226,112 @@ def DetachServiceAccount(self) -> None:
...
+
+    Returns:
+      List[Any]: the normalised l4config"""
+    normalised_l4config = []
+    for config in l4config:
+      normalised = {}
+      if 'ipProtocol' in config:
+        normalised['ip_protocol'] = config['ipProtocol']
+      else:
+        normalised['ip_protocol'] = config['IPProtocol']
```
Comment (giovannt0) [inline libcloudforensics/providers/gcp/internal/compute.py:1244]: If `ipProtocol` is not in `config`, is it always guaranteed that `IPProtocol` is?

**Review by giovannt0 — CHANGES_REQUESTED** [review]

File: libcloudforensics/providers/gcp/internal/compute.py (hunk 4 above)
Comment (jonathan-greig) [reply libcloudforensics/providers/gcp/internal/compute.py:1244]: Yeah according to the documentation each "L4config" will have either an `ipProtocol` if it is a firewall policy or `IPProtocol` if it is a standard firewall rule. Not sure why they are named differently :smiley: . `elif` is probably more suitable here though will change.

**Review by giovannt0 — APPROVED** [review]

**Pull request merged**